### PR TITLE
[Bugfix] Fix guided_decode LogitsProcessor FSM_state missing error

### LINKED ISSUE #7827: [Bug]: tool_calls parsing error with CPU
### Your current environment

<details>
<summary>The output of `python collect_env.py`</summary>

```text
Your output of `python collect_env.py` here
```

</details>

### 🐛 Describe the bug

```python
from openai import OpenAI

if __name__ == "__main__":
    
    model = "mistralai/Mistral-7B-Instruct-v0.3"
    llm = OpenAI(base_url="http://100.83.111.250:8000/v1", api_key="EMPTY")

    input = [{"role": "user", "content": "What's the weather like in Boston today?"}]
    tools = [{
        "type": "function",
        "function": {
            "name": "get_current_weather",
            "description": "Get the current weather in a given location",
            "parameters": {
                "type": "object",
                "properties": {
                "location": {
                    "type": "string",
                    "description": "The city and state, e.g. San Francisco, CA",
                },
                "unit": {"type": "string", "enum": ["celsius", "fahrenheit"]},
                },
                "required": ["location"],
            },
        }
    }]                   
    tool_choice = {'function': {'name': 'get_current_weather'}} 
    
    output = llm.chat.completions.create(
        model=model,
        messages=input,
        tools=tools,
        tool_choice=tool_choice,
        stream=False,
    )
    if content := output.choices[0].message.content:
        print('content:', content)
    elif tool_calls := output.choices[0].message.tool_calls:
        print('tool_calls:', tool_calls)
```

received output:
``` bash
python test_openai_sdk.py 
tool_calls: [ChatCompletionMessageToolCall(id='chatcmpl-tool-a081c6fb632c4d7bbfd3ed00ac5a4399', function=Function(arguments='{ ', name='get_current_weather'), type='function')]
```

expected output
```bash
python test_openai_sdk.py 
tool_calls: [ChatCompletionMessageToolCall(id='chatcmpl-tool-b58a97994c5d4de7bc817a36dcf5e95d', function=Function(arguments='{ "location": "Boston, MA", "unit": "celsius" }', name='get_current_weather'), type='function')]
```

### Before submitting a new issue...

- [X] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

### PR DESCRIPTION
**Pull request by xuechendi**
FILL IN THE PR DESCRIPTION HERE

FIX https://github.com/vllm-project/vllm/issues/7827

* What is the bug

For tool_calls handling, tested with [ISSUE_7827](https://github.com/vllm-project/vllm/issues/7827), returned tool_calls is not in correct format.

After investigating the codes, error is caused by _fsm_states object being refreshed everytime with new model_execute. While the logit_processor initial purpose is using a consistent _fsm_state crossing requests.

* What is the fix

Pass sampling_metadata_cache in model_execute, and use existing sampling_param in the cache.

* How it tested

Tested with same script indicated in [ISSUE_7827](https://github.com/vllm-project/vllm/issues/7827)

Before fixing, output is 
```python test_openai_sdk.py 
tool_calls: [ChatCompletionMessageToolCall(id='chatcmpl-tool-a081c6fb632c4d7bbfd3ed00ac5a4399', function=Function(arguments='{ ', name='get_current_weather'), type='function')]
```

After fixing, output is 
```python test_openai_sdk.py 
tool_calls: [ChatCompletionMessageToolCall(id='chatcmpl-tool-b58a97994c5d4de7bc817a36dcf5e95d', function=Function(arguments='{ "location": "Boston, MA", "unit": "celsius" }', name='get_current_weather'), type='function')]
```

**BEFORE SUBMITTING, PLEASE READ THE CHECKLIST BELOW AND FILL IN THE DESCRIPTION ABOVE**

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

**Comment by xuechendi** [comment]
@zhouyuan, may you help me to review this PR and @ who may as reviewers. Thanks

File: vllm/model_executor/sampling_metadata.py [hunk 1]
```diff
@@ -249,16 +249,20 @@ def _prepare_seq_groups(
...
         seq_ids = seq_group_metadata.seq_data.keys()
 
+        sampling_params = seq_group_metadata.sampling_params
         if cache is not None:
             sample_obj = cache.get_cached_seq_group_to_sample(len(seq_ids))
 
+            if sample_obj.sampling_params is not None:
+                sampling_params.logits_processors = \
+                    sample_obj.sampling_params.logits_processors
+
```
Comment (bigPYJ1151) [inline vllm/model_executor/sampling_metadata.py:259]: Does the GPU backend have the problem? In general, the ```cache``` is only used for reducing Python object allocation overhead, the content of the cache should be overwritten rather than reusing. Maybe the ```logits_processors``` should be filled at some other place.

Comment (xuechendi) [reply vllm/model_executor/sampling_metadata.py:259]: Thanks, I realized that this Bug only occurs when invoking async_execute_model, the main reason is that the sampling_param gets copied when submitting to mp.Queue.

I agree that putting under SamplingMetaCache is not a good way, still trying to find a better way to fix this error.

File: vllm/worker/worker_base.py [hunk 2]
```diff
@@ -295,6 +295,22 @@ def execute_model(
...
+
+        if execute_model_req:
+            for seq_group in execute_model_req.seq_group_metadata_list:
+                if seq_group.request_id not in self.global_logits_processors:
+                    self.global_logits_processors[seq_group.request_id] = \
+                        seq_group.sampling_params.logits_processors
+                else:
+                    seq_group.sampling_params.logits_processors = \
+                        self.global_logits_processors[seq_group.request_id]
+
```
Comment (xuechendi) [inline vllm/worker/worker_base.py:313]: @bigPYJ1151 , please check if this new solution is OK to you?
